docs/storage-layout-reorg-v0.10.md

```diff
@@ -0,0 +1,102 @@
+
+## Overview
+
+Ark v0.10 includes breaking changes to where data is stored in your object storage bucket. You'll need to run a [one-time migration procedure](#upgrading-to-v0.10)
+if you're upgrading from prior versions of Ark.
+
+## Details
+
+Prior to v0.10, Ark stored data in an object storage bucket using the following structure:
+
+```
+<your-bucket>/
+    backup-1/
+        ark-backup.json
+        backup-1.tar.gz
+        backup-1-logs.gz
+        restore-of-backup-1-logs.gz
+        restore-of-backup-1-results.gz
+    backup-2/
+        ark-backup.json
+        backup-2.tar.gz
+        backup-2-logs.gz
+        restore-of-backup-2-logs.gz
+        restore-of-backup-2-results.gz
+    ...
+```
+
+As of v0.10, we've reorganized this layout to provide a cleaner and more extensible directory structure. The new layout looks like:
+
+```
+<your-bucket>[/<your-prefix>]/
+    backups/
+        backup-1/
+            ark-backup.json
+            backup-1.tar.gz
+            backup-1-logs.gz
+        backup-2/
+            ark-backup.json
+            backup-2.tar.gz
+            backup-2-logs.gz
+        ...
+    restores/
+        restore-of-backup-1/
+            restore-of-backup-1-logs.gz
+            restore-of-backup-1-results.gz
+        restore-of-backup-2/
+            restore-of-backup-2-logs.gz
+            restore-of-backup-2-results.gz
+        ...
+    ...
+```
+
+## Upgrading to v0.10
+
+Before upgrading to v0.10, you'll need to run a one-time upgrade script to rearrange the contents of your existing Ark bucket(s) to be compatible with
+the new layout.
+
+Please note that the following scripts **will not** migrate existing restore logs/results into the new `restores/` subdirectory. This means that they
+will not be accessible using `ark restore describe` or `ark restore logs`. They *will* remain in the relevant backup's subdirectory so they are manually
+accessible, and will eventually be garbage-collected along with the backup. We've taken this approach in order to keep the migration scripts simple 
+and less error-prone.
+
+### rclone-Based Script
+
+This script uses [rclone][1], which you can download and install following the instructions [here][2]. 
+Please read through the script carefully before starting and execute it step-by-step.
+
+```bash
+PROVIDER=<azure|gcs|s3>
+ARK_BUCKET=<your-ark-bucket>
+ARK_TEMP_MIGRATION_BUCKET=<a-temp-bucket-for-migration>
+
+# 1. This is an interactive step that configures rclone to be 
+#    able to access your storage provider. Follow the instructions:
+rclone config
+
+# 2. Create a temporary bucket to be used as a backup of your 
+#    current Ark bucket's contents:
+rclone mkdir ${PROVIDER}:${ARK_TEMP_MIGRATION_BUCKET}
+
+# 3. Do a full copy of the contents of your Ark bucket into the 
+#    temporary bucket:
+rclone copy ${PROVIDER}:${ARK_BUCKET} ${PROVIDER}:${ARK_TEMP_MIGRATION_BUCKET}
+
+# 4. STOP and manually verify that the temporary bucket contains
```

Comment:
Past experience has led me to be a little wary of steps like this, as they'll often be skipped or skimmed since they're not executable lines.

Without adding too much logic, maybe we could diff the output of `rclone ls` on each bucket?

Reply:
Let me see what I can come up with

Reply:
Might have to play with the options a little bit, but this is kind of what I had in mind.

```
% diff <(rclone lsf -R ${PROVIDER}:${ARK_TEMP_MIGRATION_BUCKET}) <(rclone lsf -R ${PROVIDER}:${ARK_BUCKET})

13a14,16
> default/ark-backup.json
> default/default-logs.gz
> default/default.tar.gz
19,21d21
< default/ark-backup.json
< default/default-logs.gz
< default/default.tar.gz
```

Used https://rclone.org/commands/rclone_ls/ for reference

Reply:
Also, here's what it looked like after I did the `rclone delete`.

```
x1c in /home/nrb
% rclone delete ${PROVIDER}:${ARK_BUCKET}

x1c in /home/nrb
% diff <(rclone lsf -R ${PROVIDER}:${ARK_TEMP_MIGRATION_BUCKET}) <(rclone lsf -R ${PROVIDER}:${ARK_BUCKET})

1,24d0
< all/
< b1/
< def/
< default/
< nginx/
< nginx/ark-backup.json
< nginx/nginx-logs.gz
< nginx/nginx.tar.gz
< nginx/restore-nginx-20180921115829-logs.gz
< nginx/restore-nginx-20180921115829-results.gz
< def/ark-backup.json
< def/def-logs.gz
< def/def.tar.gz
< b1/ark-backup.json
< b1/b1-logs.gz
< b1/b1.tar.gz
< b1/restore-b1-20180907170648-logs.gz
< b1/restore-b1-20180907170648-results.gz
< default/ark-backup.json
< default/default-logs.gz
< default/default.tar.gz
< all/all-logs.gz
< all/all.tar.gz
< all/ark-backup.json
```

Reply:
Ooh, found an even better option - https://rclone.org/commands/rclone_check/